File: app/src/main/java/com/starsep/sokoban/view/GameView.java
package com.starsep.sokoban.view;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.starsep.sokoban.mvc.GameController;
import com.starsep.sokoban.R;
import com.starsep.sokoban.res.Textures;
import com.starsep.sokoban.mvc.ViewEventsListener;
import com.starsep.sokoban.mvc.GameModel;
import com.starsep.sokoban.gamelogic.HighScore;
import com.starsep.sokoban.gamelogic.Level;
import com.starsep.sokoban.gamelogic.Position;
import com.starsep.sokoban.gamelogic.Tile;

public class GameView extends View implements ViewEventsListener {
	private Rect dimension;
	private int size; // size of tile
	private Paint textPaint;
	private Position screenDelta;
	private GameController gameController;
	private GameModel gameModel;
	private Dialog winDialog;

	public GameView(Context context, AttributeSet attributeSet) {
		super(context, attributeSet);

		screenDelta = new Position(0, 0);

		Textures.init(getContext());

		int size = Math.min(getWidth(), getHeight()) / 10;
		dimension = new Rect(0, 0, size, size);
		textPaint = new Paint();
		textPaint.setColor(Color.BLACK);
	}

	private void setDrawingDimension(int x, int y) {
		dimension.set(screenDelta.x + x * size, screenDelta.y + y * size,
				screenDelta.x + (x + 1) * size, screenDelta.y + (y + 1) * size);
	}

	private void drawHero(@NonNull Canvas canvas) {
		setDrawingDimension(gameModel.player().x, gameModel.player().y);
		canvas.drawBitmap(gameModel.lastMove().heroTexture(), null, dimension, null);
	}

	private void drawTiles(@NonNull Canvas canvas) {
		Level level = gameModel.level();
		for (int y = 0; y < level.height(); y++) {
			for (int x = 0; x < level.width(); x++) {
				setDrawingDimension(x, y);
				if (!Tile.isGrass(level.tile(y, x))) {
					canvas.drawBitmap(level.texture(y, x), null, dimension, null);
				}
			}
		}
	}

	private void updateSize() {
		Level level = gameModel.level();
		int newSize = Math.min(getWidth() / level.width(), getHeight() / level.height());
		if (newSize != size) {
			size = newSize;
			textPaint.setTextSize(size);
			screenDelta.x = (getWidth() - level.width() * size) / 2;
			screenDelta.y = (getHeight() - level.height() * size) / 2;
		}
	}

	@Override
	protected void onDraw(@NonNull Canvas canvas) {
		super.onDraw(canvas);
		if (gameModel == null) {
			return;
		}
		updateSize();
		drawTiles(canvas);
		drawHero(canvas);
	}

	@Override
	public void showWinDialog(int levelNumber, @NonNull HighScore levelStats, @Nullable HighScore highScore) {
		if (highScore == null) {
			highScore = levelStats;
		}
		int minutes = levelStats.time / 60;
		int seconds = levelStats.time % 60;
		int minutesBest = highScore.time / 60;
		int secondsBest = highScore.time % 60;
		String msg = String.format(getResources().getString(R.string.win_msg),
				levelStats.moves, highScore.moves, levelStats.pushes, highScore.pushes,
				minutes, seconds, minutesBest, secondsBest);
		winDialog = new AlertDialog.Builder(getContext())
				.setTitle(String.format(getResources().getString(R.string.win_title), levelNumber))
				.setMessage(msg)
				.setPositiveButton(getResources().getString(R.string.win_positive), new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						gameModel.nextLevel();
					}
				})
				.setNegativeButton(getResources().getString(R.string.win_negative), new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						gameModel.repeatLevel();
					}
				})
				.setOnCancelListener(new DialogInterface.OnCancelListener() {
					@Override
					public void onCancel(DialogInterface dialog) {
						gameModel.nextLevel();
					}
				})
				.setIcon(android.R.drawable.ic_dialog_info)
				.create();
		winDialog.show();
	}

	@Override
	protected void onDetachedFromWindow() {
		super.onDetachedFromWindow();
		if (winDialog != null) {
			winDialog.dismiss();
		}
	}

	@Override
	public void onUpdate() {
		invalidate();
		if (gameController != null) {
			gameController.onStatsChanged();
		}
	}

	public void setGameController(@NonNull GameController controller) {
		gameController = controller;
		onUpdate();
	}

	public void setGameModel(@NonNull GameModel model) {
		gameModel = model;
	}
}

File: app/src/main/java/com/starsep/sokoban/view/GameView.kt
package com.starsep.sokoban.view

import android.app.AlertDialog
import android.app.Dialog
import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.support.v4.app.FragmentActivity
import android.util.AttributeSet
import android.util.Log
import android.view.View

import com.starsep.sokoban.R
import com.starsep.sokoban.Sokoban
import com.starsep.sokoban.gamelogic.HighScore
import com.starsep.sokoban.gamelogic.level.ImmutableLevel
import com.starsep.sokoban.gamelogic.Position
import com.starsep.sokoban.gamelogic.Tile
import com.starsep.sokoban.gamelogic.level.Level
import com.starsep.sokoban.model.GameModel
import com.starsep.sokoban.res.Textures

class GameView(context: Context, attributeSet: AttributeSet) : View(context, attributeSet) {
    private val dimension: Rect
    private val textPaint: Paint = Paint().apply { color = Color.BLACK }
    private var screenDelta: Position = Position(0, 0)
    private var size: Int = 0 // size of tile
    private val activity: FragmentActivity = context as FragmentActivity
    private val gameModel: GameModel by lazy {
        ViewModelProviders.of(activity).get(GameModel::class.java)
    }
    private var winDialog: Dialog? = null

    init {
        Textures.init(context)

        size = Math.min(width, height) / 10
        dimension = Rect(0, 0, size, size)

        gameModel.levelLive.observe(activity, Observer<Level> { level ->
            level?.let {
                val newSize = Math.min(width / level.width, height / level.height())
                if (newSize != size) {
                    size = newSize
                    textPaint.textSize = size.toFloat()
                    screenDelta = Position(
                            (height - level.height() * size) / 2,
                            (width - level.width * size) / 2
                    )
                }
            }
        })
    }

    private fun setDrawingDimension(x: Int, y: Int) {
        dimension.set(screenDelta.x + x * size, screenDelta.y + y * size,
                screenDelta.x + (x + 1) * size, screenDelta.y + (y + 1) * size)
    }

    private fun drawHero(canvas: Canvas) {
        setDrawingDimension(gameModel.player().x, gameModel.player().y)
        val bitmap = gameModel.lastMove()?.heroTexture() ?: Textures.heroDown()
        canvas.drawBitmap(bitmap, null, dimension, null)
    }

    private fun drawTiles(canvas: Canvas) {
        val level = gameModel.level()
        Log.d(Sokoban.TAG, level.toString())
        for (y in 0 until level.height()) {
            for (x in 0 until level.width) {
                setDrawingDimension(x, y)
                Log.d(Sokoban.TAG, x.toString() + " asd " + y.toString())
                if (!Tile.isGrass(level.tile(Position(y, x)))) {
                    canvas.drawBitmap(level.texture(Position(y, x)), null, dimension, null)
                }
            }
        }
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawTiles(canvas)
        drawHero(canvas)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        winDialog?.dismiss()
    }

    fun showWinDialog(levelNumber: Int, stats: HighScore, highScore: HighScore) {
        val minutes = stats.time / 60
        val seconds = stats.time % 60
        val minutesBest = highScore.time / 60
        val secondsBest = highScore.time % 60
        val msg = String.format(resources.getString(R.string.win_msg),
                stats.moves, highScore.moves, stats.pushes, highScore.pushes,
                minutes, seconds, minutesBest, secondsBest)
        winDialog = AlertDialog.Builder(context)
                .setTitle(String.format(resources.getString(R.string.win_title), levelNumber))
                .setMessage(msg)
                .setPositiveButton(resources.getString(R.string.win_positive)) { _, _ -> gameModel.nextLevel(context) }
                .setNegativeButton(resources.getString(R.string.win_negative)) { _, _ -> gameModel.repeatLevel(context) }
                .setOnCancelListener { gameModel.nextLevel(context) }
                .setIcon(android.R.drawable.ic_dialog_info)
                .create()
        winDialog?.show()
    }
}
